### ISSUE TITLE
Doesn't work on Android

### ISSUE THREAD

**Issue by R4md4c** [benas/random-beans#62]
Hello,

I was interested to try this library in my Android project. 

What I found is that it works with no problems when I use it in my unit tests that runs on the JVM directly. But it crashes with a weird error when tried it in my instrumentation tests on a real device.

Here is what I tried

``` java
final Populator populator = PopulatorBuilder.aNewPopulatorBuilder().build();
List<MyModel> catalogArticleList = populator.populateBeans(MyModel.class, 20);
```

And the weird error is

```
A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x2 in tid 4098 (roidJUnitRunner)
```

I suspect it is related to Android VM, and the way it execute the byte code that is generated from this library.

I know that this library was written primarily to be used in Backends. But it would be really great to port it to Android.

Cheers

**Comment by benas** [comment]
> it would be really great to port it to Android.

Sure! Random Beans should be Android compatible.
I'm not an Android expert .. but the library uses the reflection API , do you think this would be the cause of the problem? Or may be one of the dependencies of Random Beans that causes the issue.

> A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x2 in tid 4098 (roidJUnitRunner)

I do confirm, that's weird :smile: Do you have some stack trace that may help?

Which version of the library are you using?

**Comment by R4md4c** [comment]
I'm using 2.0.0-SNAPSHOT.
Here is what I got, I don't think it is a helpful stacktrace.

```
02-05 12:00:07.305 31771-31771/? I/DEBUG: pid: 4848, tid: 4898, name: roidJUnitRunner  >>> my.application <<<
02-05 12:00:07.305 31771-31771/? I/DEBUG: signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x2
02-05 12:00:07.325 31771-31771/? I/DEBUG:     r0 00000000  r1 485c37b5  r2 b6f6ce04  r3 00000001
02-05 12:00:07.325 31771-31771/? I/DEBUG:     r4 13060000  r5 ffffffff  r6 00000000  r7 af080a80
02-05 12:00:07.325 31771-31771/? I/DEBUG:     r8 af077c00  r9 b4e857fc  sl 00080011  fp 00000002
02-05 12:00:07.325 31771-31771/? I/DEBUG:     ip b3bb78d8  sp b3bb7920  lr 001ecd02  pc b4c10f80  cpsr 60070030
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #00 pc 00110f80  /system/lib/libart.so (art::gc::allocator::RosAlloc::AllocFromRun(art::Thread*, unsigned int, unsigned int*)+195)
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #01 pc 000f654f  /system/lib/libart.so (art::mirror::Class::AllocObject(art::Thread*)+774)
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #02 pc 001d00c3  /system/lib/libart.so (art::JNI::NewObjectV(_JNIEnv*, _jclass*, _jmethodID*, std::__va_list)+322)
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #03 pc 000b7141  /system/lib/libart.so (art::CheckJNI::NewObjectV(_JNIEnv*, _jclass*, _jmethodID*, std::__va_list)+64)
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #04 pc 0000ec29  /system/lib/libjavacore.so
02-05 12:00:07.325 31771-31771/? I/DEBUG:     #05 pc 00047405  /system/framework/arm/boot.oat
```

**Comment by R4md4c** [comment]
I speculate that it could be related to how Mockito works in Android, as it doesn't work out of the box, you have to include DexMaker with it, so that it could work with the Android VM. [1]

Perhaps Random Beans needs something like a DexMaker proxy?

[1] https://github.com/crittercism/dexmaker/tree/master/mockito/src/main/java/com/android/dx/mockito

**Comment by benas** [comment]
Hi,

I'm not sure I understand, Random Beans does not have a compile-time dependency to mockito. So how this could be related to the way on how does mockito works in android (and the workaround needed)?

**Comment by R4md4c** [comment]
Hi,

Sorry for not being clear enough, but what I was trying to say is by bringing an analogy to Mockito. As without DexMaker it doesn't work on Android Dalvik VM, we have to use DexMaker to make it compatible with the Android Dalvik VM. 

So I thought that we might have the same problem with RandomBeans, and we need to use DexMaker to make it compatible with Android Dalvik VM.

I'm just speculating, and of course I'm not sure about weather I'm saying is right or wrong.

**Comment by benas** [comment]
Hi @R4md4c ,

In order to reproduce the issue, I'll setup an Android development environment and use the library on a Dalvik VM. Can you please give me all details so I can have the same setup as you?

* Android SDK version
* Java version
* OS version
* The test case with the bean being populated

Thank you in advance
Mahmoud

**Comment by R4md4c** [comment]
Hi Mahmoud,

I've prepared a test project for you with a failing test case. [1]

All you have to do is that you setup an emulator, and run

`./gradlew connectedAndroidTest`

And then it will download some dependecies and execute the test case on the emulator that you've started 

I'm using Android SDK Version 23, Java Version 8. 

[1] https://www.dropbox.com/s/75t9ru23j402dwc/RandomBeans.zip?dl=0

**Comment by benas** [comment]
Oh that's so kind! thank you :smile: 

I'll give it a try and keep you informed.
I've planned to fix this issue in version 3. Currently I'm preparing to release v2 shortly.

Regards
Mahmoud

**Comment by R4md4c** [comment]
Ok Great, Let me know I might help if you managed to find the problem.

**Comment by PascalSchumacher** [comment]
I'm not android developer, but I was able to create a hello world android application with the android sdk use random-bean with androidtest:

```
androidTestCompile('io.github.benas:random-beans:2.0.0') {
        exclude module: 'commons-beanutils'
        exclude module: 'commons-math3'
 }
```

```java
@RunWith(AndroidJUnit4.class)
public class RandomBeansTest {

    @Test
    public void testRandomBeans(){
        EnhancedRandom enhancedRandom = EnhancedRandomBuilder.aNewEnhancedRandomBuilder().build();
        Date date = enhancedRandom.nextObject(Date.class);

        Assert.assertNotNull(date);
    }
}
```

```
....
:app:connectedAndroidTest

BUILD SUCCESSFUL
```

@R4md4c I could not get your example to run, because I was getting:  `com.android.dex.DexException: Multiple dex files define Ljavax/annotation/CheckForNull;`

**Comment by R4md4c** [comment]
@PascalSchumacher The idea is to start an Activity while executing your @Test method. 
The main difference between my code and yours is this line
``` 
@Rule
public ActivityTestRule activityTestRule = new ActivityTestRule<>(MainActivity.class);
```

This Rule makes JUnit to start an Activity (which is a screen in Android), and close it once this @Test method is finished.

**Comment by PascalSchumacher** [comment]
@R4md4c Thanks for the info. I added the rule and it still works. 

I have pushed the complete example to  https://github.com/PascalSchumacher/RandomBeansAndroidTest

**Comment by benas** [comment]
Hi @R4md4c 

Your [first test](https://github.com/benas/random-beans/issues/62#issuecomment-180297555) was with version 2.0.0-SNAPSHOT right? Have you tried to use the release version 2.0.0 ? Looks like this is the difference between your project and the one shared by Pascal (Thank you Pascal for sharing your test project).

It's also relevant to try with the latest 3.0.0-SNAPSHOT. I'll test on my side with both versions too.

Kind regards
Mahmoud

**Comment by benas** [comment]
Hi @R4md4c 

I was able to setup an run your project. I have the following error:

```java
UNEXPECTED TOP-LEVEL EXCEPTION:
com.android.dex.DexException: Multiple dex files define Ljavax/annotation/CheckForNull;
	at com.android.dx.merge.DexMerger.readSortableTypes(DexMerger.java:579)
	at com.android.dx.merge.DexMerger.getSortedTypes(DexMerger.java:535)
	at com.android.dx.merge.DexMerger.mergeClassDefs(DexMerger.java:517)
	at com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:164)
	at com.android.dx.merge.DexMerger.merge(DexMerger.java:188)
	at com.android.dx.command.dexer.Main.mergeLibraryDexBuffers(Main.java:504)
	at com.android.dx.command.dexer.Main.runMonoDex(Main.java:334)
	at com.android.dx.command.dexer.Main.run(Main.java:277)
	at com.android.dx.command.dexer.Main.main(Main.java:245)
	at com.android.dx.command.Main.main(Main.java:106)
```

The problem is, even *without* executing Random Beans code, I get the same error:

```java

//import io.github.benas.randombeans.EnhancedRandomBuilder;
//import io.github.benas.randombeans.api.EnhancedRandom;

/**
 * Created by r4md4c.
 */
@RunWith(AndroidJUnit4.class)
public class MainActivityTest {

    @Rule
    public ActivityTestRule activityTestRule = new ActivityTestRule<>(MainActivity.class);

    @Test
    public void testRandomBeans() throws Exception {
        /*EnhancedRandom enhancedRandom = EnhancedRandomBuilder.aNewEnhancedRandomBuilder().build();
        DummyModel dummyModel = enhancedRandom.nextObject(DummyModel.class);

        Assert.assertNotNull(dummyModel.getId());*/
        System.out.println("ok");
    }
}
```

I used:

* random beans: 2.0.0
* android sdk: 23
* java: 8
* win7 64 bits

Looks like just having random beans in the build path (without calling the library code) causes the error, because when I removed the dependency from gradle (`//    androidTestCompile 'io.github.benas:random-beans:2.0.0'`), everything works fine. hmmm..

I'll continue my investigation and keep you informed.

Kind regards
Mahmoud

**Comment by R4md4c** [comment]
Hi @benas  and @PascalSchumacher 

This is my Gradle configuration that works with random-beans 2.0.0, but still it crashes for an unknown error.

```
apply plugin: 'com.android.application'

repositories {
    maven {
        url 'https://oss.sonatype.org/content/repositories/snapshots'
    }
}

android {
    compileSdkVersion 23
    buildToolsVersion "23.0.2"

    defaultConfig {
        applicationId "com.example.randombeans"
        minSdkVersion 15
        targetSdkVersion 23
        versionCode 1
        versionName "1.0"
        multiDexEnabled true
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        debug {
            minifyEnabled false
        }
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }

    packagingOptions {
        exclude 'META-INF/LICENSE.txt'
        exclude 'META-INF/NOTICE.txt'
        exclude 'META-INF/maven/com.google.guava/guava/pom.properties'
        exclude 'META-INF/maven/com.google.guava/guava/pom.xml'
    }
}

dependencies {
    compile 'com.android.support:appcompat-v7:23.2.1'
    compile 'com.android.support:design:23.2.1'

    androidTestCompile 'junit:junit:4.12'
    androidTestCompile('io.github.benas:random-beans:2.0.0') {
        exclude group: 'com.google.code.findbugs'
    }
    androidTestCompile("com.android.support.test.espresso:espresso-core:2.2.1") {
        exclude module: 'support-annotations'
    }
    androidTestCompile('com.android.support.test:runner:0.4.1') {
        exclude module: 'support-annotations'
    }
}
```

**Comment by benas** [comment]
I was suspecting that one of random beans dependencies would cause the problem. so I tried to include them one by one (there are only two: objenesis and fast-classpath-scanner).

None of them seems to cause the problem. It is only when I add random beans dependency to the project that it fails. According to the error, Multiple dex files are defined for `javax/annotation/CheckForNull`. Not sure how random beans could brings this class to the table..

I'll continue to analyze this problem and keep you informed.

**Comment by korniltsev** [comment]
I've tried to use the lib on android and found two problems so far.
1. inside of `org.reflections.util.ClasspathHelper#forJavaClassPath ``System.getProperty("java.class.path")` returns "." on Android. So the lib fails to analyze the classpath (crashes VM on pre lollipop (API < 21 has Dalvik runtime) and silently fails to analyze the classpath on Lollipop+ (API >= 21 has Art runtime, maybe). So it tries to create class file from every file in entire file system and crashes(API < 21) VM trying to create class from "/dev/ptmx".
![2016-12-07 11 05 58](https://cloud.githubusercontent.com/assets/331773/20960020/4f41abbe-bc6f-11e6-81f3-9adaa2cc94c2.png)

2. The second one is also runtime related. Since the first problem is not fatal on Api 21 I managed to make the lib work using this trick
``` EnhancedRandom enhancedRandom = 
EnhancedRandomBuilder.aNewEnhancedRandomBuilder()
              .exclude(field().named("shadow$_klass_").ofType(Class.class).inClass(Object.class).get())
                .build();
```
which is Android Art runtime specific field.
Note: since `org.reflections.Reflections` fails to scan the classpath properly, it means `io.github.benas.randombeans.util.ReflectionUtils#searchForPublicConcreteSubTypesOf` works incorrect on android.
It would be nice to have this exclude by default when running on Android.

**Comment by korniltsev** [comment]
There is also a "shadow$_monitor_" an Art runtime

**Comment by korniltsev** [comment]
since v3 branch is not suitable for android, and v2 is not maintained, this should be closed

**Comment by benas** [comment]
We didn't gave up on this yet :muscle:  :smile: 

> Note: since org.reflections.Reflections fails to scan the classpath properly, it means io.github.benas.randombeans.util.ReflectionUtils#searchForPublicConcreteSubTypesOf works incorrect on android. It would be nice to have this exclude by default when running on Android.

this is an optional dependency, it is not included until `scanClasspathForConcreteTypes` is activated. Have you tried to use random beans on android without activating this parameter?

**Comment by korniltsev** [comment]
My comment was about v2 branch. And it crashes during class initialization since a Reflections instance is in static final field.
 master branch is not suitable for android because of java 8 and stream api

**Comment by benas** [comment]
> My comment was about v2 branch. And it crashes during class initialization since a Reflections instance is in static final field.

Oh yes sorry about that. Indeed in v2, `org.reflections` was not an optional dependency.

**Comment by benas** [comment]
The following is a good news to continue working on this issue:

https://android-developers.googleblog.com/2017/03/future-of-java-8-language-feature.html

Looking forward to it!

**Comment by benas** [comment]
Hi,

I updated the [known-limitations](https://github.com/benas/random-beans/wiki/known-limitations#android-compatibility) section to mention that RB is not compatible with Android.

Unfortunately, I am not able to put too much effort in making it Android compatible. Thank you for your comprehension. I'm closing this issue.

Best regards,
Mahmoud
